lib/appsignal/helpers/instrumentation.rb

```diff
@@ -5,6 +5,144 @@ module Helpers
+      #
+      # This is a helper to create an AppSignal transaction, track any errors
+      # that may occur and complete the transaction.
+      #
+      # This helper is recommended to be used in Ruby scripts and parts of an
+      # app not already instrumented by AppSignal's automatic instrumentations.
+      #
+      # Use this helper in combination with our {.instrument} helper to track
+      # instrumentation events.
+      #
+      # If AppSignal is not active ({Appsignal.active?}) it will still execute
+      # the block, but not create a transaction for it.
+      #
+      # @example Instrument a block of code
+      #   Appsignal.monitor(
+      #     :namespace => "my_namespace",
+      #     :action => "MyClass#my_method"
+      #   ) do
+      #     # Some code
+      #   end
+      #
+      # @example Instrument a block of code with an instrumentation event
+      #   Appsignal.monitor(
+      #     :namespace => "my_namespace",
+      #     :action => "MyClass#my_method"
+      #   ) do
+      #     Appsignal.instrument("some_event.some_group") do
+      #       # Some code
+      #     end
+      #   end
+      #
+      # @example Set custom metadata on the transaction
+      #   Appsignal.monitor(
+      #     :namespace => "my_namespace",
+      #     :action => "MyClass#my_method"
+      #   ) do
+      #     # Some code
+      #
+      #     Appsignal.set_tags(:tag1 => "value1", :tag2 => "value2")
+      #     Appsignal.set_params(:param1 => "value1", :param2 => "value2")
+      #   end
+      #
+      # @example Call monitor within monitor (not recommended)
+      #   Appsignal.monitor(
+      #     :namespace => "my_namespace",
+      #     :action => "MyClass#my_method"
+      #   ) do
+      #     # This will _not_ update the namespace and action name
+      #     Appsignal.monitor(
+      #       :namespace => "my_other_namespace",
+      #       :action => "MyOtherClass#my_other_method"
+      #     ) do
+      #       # Some code
+      #
+      #       # The reported namespace will be "my_namespace"
+      #       # The reported action will be "MyClass#my_method"
+      #     end
+      #   end
+      #
+      # @param namespace [String/Symbol] The namespace to set on the new
+      #   transaction.
+      #   Defaults to the 'web' namespace.
+      #   This will update the already active transaction's namespace if
+      #   {.monitor} is called when another transaction is already active.
+      # @param action [String/Symbol] The action name for the transaction.
+      #   This will update the already active transaction's action if
+      #   {.monitor} is called when another transaction is already active.
+      # @yield The block to monitor.
+      # @raise [Exception] Any exception that occurs within the given block is
+      #   re-raised by this method.
+      # @return [Object] The value of the given block is returned.
+      # @since 3.11.0
+      def monitor(
+        namespace: nil,
+        action: nil
```

Comment:
These shouldn't be optional, right? Or at least not `action`. If no action name is set, it won't create a valid transaction.

Reply:
Right, I see the intent! I worry that it will lead to customers not setting the action or using `set_action` and being confused about their transaction not showing up.

I would suggest making the argument not-optional-but-nillable: that is, you _have_ to specify `action:` when calling `monitor`, but you can specify `action: nil` (or set a placeholder) and then override it later on with `Appsignal.set_action`.

We could also leave it optional and fix this through good documentation, but I think, when possible, we should prefer API design that hints the customer into making the right choice, over documenting what the right choice is.

Reply:
That would make the intent a lot more explicit. I like it :)

```ruby
Appsignal.monitor(action: nil) do
  # Do stuff
  Appsignal.set_action("My Action")
end
```

Maybe a `:set_later` placeholder value would make it even more explicit. I would make it work with both the placeholder and the nil value.

Reply:
@unflxw I've updated the helper with your suggestion. Please review :)